client/dashboard/components/settings/general/store-address.tsx

```diff
@@ -130,98 +189,101 @@ export function getCountryStateOptions() {
-			// Clear form.
+			// Clear form
 			isAutofillChange.current = true;
 			setValue( 'countryState', '' );
+			return;
 		}
-		let filteredOptions = [];
 		const countrySearch = new RegExp(
 			escapeRegExp( autofillCountry ),
 			'i'
 		);
-		const stateSearch = new RegExp(
-			escapeRegExp( autofillState.replace( /\s/g, '' ) ) + '$', // Always match the end of string for region.
-			'i'
-		);
-		if ( autofillState.length || autofillCountry.length ) {
-			filteredOptions = options.filter( ( option ) =>
-				( autofillCountry.length ? countrySearch : stateSearch ).test(
-					option.label
-				)
-			);
-		}
+		const isCountryAbbreviation = autofillCountry.length < 3;
+		const isStateAbbreviation =
+			autofillState.length < 3 && !! autofillState.match( /^[\w]+$/ );
+		let filteredOptions = [];
+
 		if ( autofillCountry.length && autofillState.length ) {
-			const isStateAbbreviation = autofillState.length < 3;
-			filteredOptions = filteredOptions.filter( ( option ) =>
-				stateSearch.test(
-					( isStateAbbreviation ? option.key : option.label )
-						.replace( '-', '' )
-						.replace( /\s/g, '' )
+			filteredOptions = options.filter( ( option ) =>
+				countrySearch.test(
+					isCountryAbbreviation ? option.key : option.label
 				)
 			);
-
-			const isCountryAbbreviation = autofillCountry.length < 3;
-			if ( filteredOptions.length > 1 ) {
-				let countryKeyOptions = [];
-				countryKeyOptions = filteredOptions.filter( ( option ) =>
-					countrySearch.test(
-						isCountryAbbreviation ? option.key : option.label
-					)
-				);
-
-				if ( countryKeyOptions.length > 0 ) {
-					filteredOptions = countryKeyOptions;
-				}
+			// no country matches so use all options for state filter.
+			if ( ! filteredOptions.length ) {
+				filteredOptions = [ ...options ];
 			}
-
 			if ( filteredOptions.length > 1 ) {
-				let stateKeyOptions = [];
-				stateKeyOptions = filteredOptions.filter( ( option ) =>
-					stateSearch.test(
-						( isStateAbbreviation ? option.key : option.label )
-							.replace( '-', '' )
-							.replace( /\s/g, '' )
+				filteredOptions = filteredOptions.filter(
+					getStateFilter(
+						isStateAbbreviation,
+						normalizeState( autofillState )
 					)
 				);
-
-				if ( stateKeyOptions.length === 1 ) {
-					filteredOptions = stateKeyOptions;
-				}
 			}
+		} else if ( autofillCountry.length ) {
+			filteredOptions = options.filter( ( option ) =>
+				countrySearch.test(
+					isCountryAbbreviation ? option.key : option.label
+				)
+			);
+		} else if ( autofillState.length ) {
+			filteredOptions = options.filter(
+				getStateFilter(
+					isStateAbbreviation,
+					normalizeState( autofillState )
+				)
+			);
 		}
-
 		if (
 			filteredOptions.length === 1 &&
 			countryState !== filteredOptions[ 0 ].key
 		) {
 			isAutofillChange.current = true;
 			setValue( 'countryState', filteredOptions[ 0 ].key );
 		}
+		// eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
Same comment as above.

Reply:
Similar here. We both read and update the `countryState`